### PR TITLE
test: add form validation test

### PR DESCRIPTION
**Pull request by hermanwikner**
### Description

This test focuses on the visibility of the validation icons displayed in each field, and whether the publish button is enabled or disabled depending on the validation of a document.

Screenshot 2021-12-09 at 15 10 28

### What to review

- The test in general and how it is authored
- Should more scenarios be tested?

### Notes for release

Add form validation test

### CODE REVIEW

**Review by ricokahler — COMMENTED** [review]
i don't think your assertions are running. i'll my findings async

File: cypress/integration/form-builder/validation.test.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+  /** Error */
+  it('when there are validation errors, the publish button is disabled', () => {
+    client
+      .patch(docId)
+      .set({errorString: INVALID_VALUE})
+      .commit()
+      .then(() => {
+        cy.get('[data-testid="action-Publish"').should('have.attr', 'disabled')
+      })
```
[the remark this answers is not in the document]
Comment (hermanwikner) [reply cypress/integration/form-builder/validation.test.js]: Thanks! I have updated the test, so this should be solved! :)

File: cypress/integration/form-builder/validation.test.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  warningString: '',
+  infoString: '',
+}
+
+/** Each fields validation rule is configured with min(5) */
+const VALID_VALUE = 'String longer than 5'
+const INVALID_VALUE = 'abc'
+
+async function patchDocument(payload) {
+  return client.patch(docId).set(payload).commit()
```
Comment (rexxars) [inline cypress/integration/form-builder/validation.test.js:19]: Cool trick: If you use async visibility here it'll return a lot quicker :)
This tells the backend to not wait for the document to be "indexed" before returning, but you'll still be able to navigate to it in the studio and such!

```suggestion
  return client.patch(docId).set(payload).commit({visibility: 'async'})
```

File: cypress/integration/form-builder/validation.test.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+describe('@sanity/form-builder: Validation', () => {
+  before(async () => {
+    await client.create(doc)
```
Comment (rexxars) [inline cypress/integration/form-builder/validation.test.js:24]:
```suggestion
    await client.create(doc, {visibility: 'async'})
```

File: cypress/integration/form-builder/validation.test.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  it('when there are no validation infos, the info icon is hidden', () => {
+    cy.wrap(patchDocument({infoString: VALID_VALUE})).then(() => {
+      cy.get('[data-testid="input-infoString"]')
+        .find('[data-testid="input-validation-icon-info"]')
+        .should('not.exist')
+    })
+  })
+
+  after(async () => {
+    await client.delete(docId)
```
Comment (rexxars) [inline cypress/integration/form-builder/validation.test.js:100]:
```suggestion
    await client.delete(docId, {visibility: 'async'})
```

**Review by rexxars — APPROVED** [review]
Looks good! Couple of suggestions on how to speed up the tests slightly - make sure to verify that the tests still works with these changes, there might be race conditions due to it being a bit faster, but hopefully not :)

File: cypress/integration/form-builder/validation.test.js (hunk 2 above)
Comment (hermanwikner) [reply cypress/integration/form-builder/validation.test.js]: Thanks! Will add 👍

**Comment by hermanwikner** [comment]
> 

Added the `{visibility: 'async'}` and the test still works. Also moved the `cy.viewport(1000, 1500)` from `beforeEach` to `before`! @rexxars

**Pull request merged**